include/ringmesh/basic/frame.h

```diff
@@ -0,0 +1,182 @@
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/basic/common.h>
+
+/*!
+ * @file Basic classes for representing frames
+ * @author Pierre Anquez
+ */
+
+namespace RINGMesh
+{
+    namespace Geometry
+    {
+        struct Plane;
+        FORWARD_DECLARATION_DIMENSION_STRUCT( Line );
+        ALIAS_2D( Line );
+    }
+}
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API FrameBase
+    {
+    public:
+        virtual ~FrameBase();
+
+        const vecn< DIMENSION >& operator[]( index_t coord ) const
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+        vecn< DIMENSION >& operator[]( index_t coord )
+        {
+            ringmesh_assert( coord < DIMENSION );
+            return axis_[coord];
+        }
+
+    protected:
+        // Default Frame aligned on space axis
+        FrameBase()
+        {
+            for( auto coord : RINGMesh::range( DIMENSION ) )
+            {
+                axis_[coord][coord] = 1.;
+            }
+        }
+
+    private:
+        std::vector< vecn< DIMENSION > > axis_{ DIMENSION,
+            vecn< DIMENSION >() };
+    };
+    ALIAS_2D_AND_3D( FrameBase );
+
+    template < index_t DIMENSION >
+    class RINGMESH_API Frame : public FrameBase< DIMENSION >
+    {
+    public:
+        Frame() = default;
+    };
+    ALIAS_2D_AND_3D( Frame );
+
+    template <>
+    class RINGMESH_API Frame< 2 > : public FrameBase< 2 >
+    {
+    public:
+        Frame() = default;
+
+        Frame( const vec2& x, const vec2& y ) : FrameBase()
+        {
+            ( *this )[0] = x;
+            ( *this )[1] = y;
+        }
+    };
+
+    template <>
+    class RINGMESH_API Frame< 3 > : public FrameBase< 3 >
+    {
+    public:
+        Frame() = default;
+
+        Frame( const vec3& x, const vec3& y, const vec3& z ) : FrameBase()
+        {
+            ( *this )[0] = x;
+            ( *this )[1] = y;
+            ( *this )[2] = z;
+        }
+    };
+
+    template < index_t DIMENSION >
+    class RINGMESH_API ReferenceFrame : public Frame< DIMENSION >
+    {
+    public:
+        ReferenceFrame() = default;
+
+        ReferenceFrame(
+            vecn< DIMENSION > frame_origin, Frame< DIMENSION > frame )
+            : Frame< DIMENSION >( std::move( frame ) ),
+              origin_( std::move( frame_origin ) )
+        {
+        }
+
+        vecn< DIMENSION > origin()
+        {
+            return origin_;
+        }
+
+    protected:
+        vecn< DIMENSION > origin_{};
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)